### PR TITLE
nixos/armagetronad: Add module with NixOS tests

### PR DESCRIPTION
**Pull request by numinit**
## Description of changes

New module, `services.armagetronad`, for hosting Armagetron dedicated servers.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by numinit** [comment]
Marking this as a draft because it'll require #263675 to be merged in order to correct the crashing (which this PR replicates)

**Comment by numinit** [comment]
Ready for review since Armagetron updates were merged.

**Review by crertel — APPROVED** [review]

File: nixos/modules/services/games/armagetronad.nix [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        default = { };
+        type = types.attrsOf (types.submodule {
+          options = {
+            enable = mkEnableOption (lib.mdDoc "armagetronad");
+            package = lib.mkPackageOptionMD pkgs "armagetronad-dedicated" {
+              example = ''
+                pkgs.armagetronad."0.2.9-sty+ct+ap".dedicated
+              '';
+              extraDescription = ''
+                Ensure that you use a derivation whose evaluation contains the path `bin/armagetronad-dedicated`.
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:41]:
```suggestion
                Ensure that you use a derivation which contains the path `bin/armagetronad-dedicated`.
```
It is technically the build not the evaluation

File: nixos/modules/services/games/armagetronad.nix [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+              '';
+            };
+            host = mkOption {
+              type = types.str;
+              default = "0.0.0.0";
+              description = lib.mdDoc "Host to listen on. Used for SERVER_IP.";
+            };
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:48]: An empty line between options would increase readability

File: nixos/modules/services/games/armagetronad.nix [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            dns = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              description = lib.mdDoc "DNS address to use for this server. Optional.";
+            };
+            openFirewall = mkOption {
+              type = types.bool;
+              default = true;
+              description = lib.mdDoc "Set to true to open a UDP port for Armagetron Advanced.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:62]:
```suggestion
              description = lib.mdDoc "Set to true to open the configured UDP port for Armagetron Advanced.";
```

File: nixos/modules/services/games/armagetronad.nix [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      (serverName: serverCfg:
+        let
+          serverId = nameToId serverName;
+          serverInfo = (
+            {
+              SERVER_IP = serverCfg.host;
+              SERVER_PORT = serverCfg.port;
+              SERVER_NAME = serverCfg.name;
+            } // (
+              if serverCfg.dns != null then { SERVER_DNS = serverCfg.dns; }
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:123]: We could use lib.optionalAttrs here

File: nixos/modules/services/games/armagetronad.nix [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+          );
+          customSettings = serverCfg.settings;
+          everytimeSettings = serverCfg.roundSettings;
+
+          serverInfoCfg = settingsFormat.generate "server_info.${serverName}.cfg" serverInfo;
+          customSettingsCfg = settingsFormat.generate "settings_custom.${serverName}.cfg" customSettings;
+          everytimeSettingsCfg = settingsFormat.generate "everytime.${serverName}.cfg" everytimeSettings;
+        in
+        {
+          "armagetronad@${serverName}" = {
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:135]: Services who contain a @ are reserved for templating. Please use a - instead.

File: nixos/modules/services/games/armagetronad.nix [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            after = [ "basic.target" "network.target" ];
+            wantedBy = [ "multi-user.target" ];
+            serviceConfig =
+              let
+                stateDirectory = "armagetronad/${serverName}";
+                serverRoot = "/var/lib/${stateDirectory}";
+                preStart = pkgs.writeShellScript "armagetronad-${serverName}-prestart.sh" ''
+                  owner="${serverId}:${serverId}"
+
+                  # Create the config directories.
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:147]: Please create directories with tmpfiles.d

File: nixos/modules/services/games/armagetronad.nix [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                  truncate -s0 "$input"
+                  chmod u+rw,g+r,o-rwx "$input"
+                  chown "$owner" "$input"
+                '';
+              in
+              {
+                Type = "simple";
+                StateDirectory = stateDirectory;
+                ExecStartPre = preStart;
+                ExecStart = "${serverCfg.package}/bin/armagetronad-dedicated --daemon --input ${serverRoot}/input --userdatadir ${serverRoot}/data --userconfigdir ${serverRoot}/settings --vardir ${serverRoot}/var --autoresourcedir ${serverRoot}/resource";
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:171]: If we use lib.getExe here, people can customize this using meta.mainProgramm

**Comment by numinit** [comment]
Thanks for the feedback @SuperSandro2000. #284626 adds meta.mainProgram to squash the warnings during the build.

**Comment by numinit** [comment]
@SuperSandro2000 I tried using systemd unit templates, but it doesn't look like they work with per-unit ExecStart being different due to potentially different packages per server. So we're stuck with `armagetronad-` vs `armagetronad@`.

File: nixos/modules/services/games/armagetronad.nix [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    systemd.services = mkMerge (mapAttrsToList
+      (serverName: serverCfg:
+        let
+          serverId = nameToId serverName;
+        in
+        {
+          "armagetronad-${serverName}" = {
+            description = "Armagetron Advanced Dedicated Server for ${serverName}";
+            wants = [ "basic.target" ];
+            after = [ "basic.target" "network.target" ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/armagetronad.nix:209]: Shouldn't this be multi-user.target, too?

**Pull request merged**